contracts/protocol/bases/PriceDiscoveryBase.sol

```diff
@@ -18,36 +18,104 @@ import { Address } from "@openzeppelin/contracts/utils/Address.sol";
 contract PriceDiscoveryBase is ProtocolBase {
     using Address for address;
 
-    IWETH9Like public immutable weth;
+    /**
+     * @notice Fulfils an order on an external contract.
+     *
+     * If the owner is price discovery contract, the protocol cannot act as an intermediary in the exchange,
+     * and sellers must use Wrapped's contract. Wrappers handle ask and bid orders in the same manner.
+     *
+     * See descriptions of `fulfilAskOrder`, `fulfilBidOrder` and handleWrapper for more details.
+     *
+     * @param _tokenId - the id of the token. Accepts whatever token is sent by price discovery contract when this value is zero.
+     * @param _offer - the fully populated BosonTypes.Offer struct
+     * @param _priceDiscovery - the fully populated BosonTypes.PriceDiscovery struct
+     * @param _buyer - the buyer's address (caller can commit on behalf of a buyer)
+     * @return actualPrice - the actual price of the order
+     */
+    function fulfilOrder(
+        uint256 _tokenId,
+        Offer storage _offer,
+        PriceDiscovery calldata _priceDiscovery,
+        address _buyer
+    ) internal returns (uint256 actualPrice) {
+        require(
+            _priceDiscovery.priceDiscoveryContract != address(0) && _priceDiscovery.conduit != address(0),
+            PRICE_DISCOVERY_CONTRACTS_NOT_SET
+        );
+
+        ProtocolLib.ProtocolLookups storage lookups = protocolLookups();
+        IBosonVoucher bosonVoucher = IBosonVoucher(lookups.cloneAddress[_offer.sellerId]);
+
+        // Set incoming voucher clone address
+        protocolStatus().incomingVoucherCloneAddress = address(bosonVoucher);
+
+        address exchangeToken = _offer.exchangeToken;
+        address owner;
+
+        if (_tokenId != 0) {
+            owner = bosonVoucher.ownerOf(_tokenId);
+        }
+
+        // Handle wrapper voucher, there is no difference between ask and bid
+        if (owner == _priceDiscovery.priceDiscoveryContract) {
+            handleWrapper(_tokenId, exchangeToken, _priceDiscovery);
+        } else if (_priceDiscovery.side == Side.Ask) {
+            return fulfilAskOrder(_tokenId, exchangeToken, _priceDiscovery, _buyer, bosonVoucher);
+        } else {
+            return fulfilBidOrder(_tokenId, exchangeToken, _priceDiscovery, bosonVoucher);
+        }
+
+        // Gets new owner
+        owner = bosonVoucher.ownerOf(_tokenId);
```

Comment:
Yes, `_tokenId` can change during `fulfilAskOrder`, but this does not affect this line, since in that case, this function returns before reaching this line.
This check happens only for `handleWrapper`, so I'm not sure why is it not a part of it. TBH, I'm not even sure if this check is needed at all, since the buyer could transfer the received voucher immediately if they want, and I don't see a reason to prevent it.
Anyway, I'll give it another thought and either put it into handleWrapper or remove it.